pkg/gameserverallocations/allocation_cache_test.go

```diff
@@ -0,0 +1,343 @@
+		features string
+	}{
+		"most allocated": {
+			// node1: 1 ready, 1 allocated, node2: 1 ready
+			list: []agonesv1.GameServer{gs1, gs2, gs3},
+			test: func(t *testing.T, list []*agonesv1.GameServer) {
+				assert.Len(t, list, 2)
+				if !assert.Equal(t, []*agonesv1.GameServer{&gs1, &gs2}, list) {
+					for _, gs := range list {
+						logrus.WithField("name", gs.Name).Info("game server")
+					}
+				}
+			},
+			features: fmt.Sprintf("%s=false", runtime.FeatureStateAllocationFilter),
+		},
+		"allocated first (StateAllocationFilter)": {
+			list:     []agonesv1.GameServer{gs1, gs2, gs3},
+			features: fmt.Sprintf("%s=true", runtime.FeatureStateAllocationFilter),
+			test: func(t *testing.T, list []*agonesv1.GameServer) {
+				require.Len(t, list, 3)
+				assert.Equal(t, "gs3", list[0].ObjectMeta.Name)
+				assert.Equal(t, "gs1", list[1].ObjectMeta.Name)
+				assert.Equal(t, "gs2", list[2].ObjectMeta.Name)
+			},
+		},
+		"nil player status (PlayerAllocationFilter)": {
+			list: []agonesv1.GameServer{gs1, gs2, gs4},
+			test: func(t *testing.T, list []*agonesv1.GameServer) {
+				require.Len(t, list, 3)
+				assert.Contains(t, []string{"gs2", "gs4"}, list[0].ObjectMeta.Name)
+				assert.Contains(t, []string{"gs2", "gs4"}, list[1].ObjectMeta.Name)
+				assert.Equal(t, &gs1, list[2])
+			},
+		},
+		"least player capacity (PlayerAllocationFilter)": {
+			list:     []agonesv1.GameServer{gs5, gs6},
+			features: fmt.Sprintf("%s=true", runtime.FeaturePlayerAllocationFilter),
+			test: func(t *testing.T, list []*agonesv1.GameServer) {
+				require.Len(t, list, 2)
+				assert.Equal(t, "gs6", list[0].ObjectMeta.Name)
+				assert.Equal(t, "gs5", list[1].ObjectMeta.Name)
+			},
+		},
+		"list ready": {
+			// node1: 1 ready, node2: 2 ready
+			list: []agonesv1.GameServer{gs1, gs2, gs4},
+			test: func(t *testing.T, list []*agonesv1.GameServer) {
+				assert.Len(t, list, 3)
+				assert.Contains(t, []string{"gs2", "gs4"}, list[0].ObjectMeta.Name)
+				assert.Contains(t, []string{"gs2", "gs4"}, list[1].ObjectMeta.Name)
+				assert.Equal(t, &gs1, list[2])
+			},
+		},
+		"lexicographical (node name)": {
+			list: []agonesv1.GameServer{gs2, gs1},
+			test: func(t *testing.T, list []*agonesv1.GameServer) {
+				assert.Len(t, list, 2)
+				if !assert.Equal(t, []*agonesv1.GameServer{&gs1, &gs2}, list) {
+					for _, gs := range list {
+						logrus.WithField("name", gs.Name).Info("game server")
+					}
+				}
+			},
+		},
+	}
+
+	for k, v := range fixtures {
+		k := k
+		v := v
+		t.Run(k, func(t *testing.T) {
+			// deliberately not resetting the Feature state, to catch any possible unknown regressions with the
+			// new feature flags
+			if v.features != "" {
+				require.NoError(t, runtime.ParseFeatures(v.features))
+			}
+
+			cache, m := newFakeAllocationCache()
+
+			gsList := v.list
+
+			m.AgonesClient.AddReactor("list", "gameservers", func(action k8stesting.Action) (bool, k8sruntime.Object, error) {
```

Comment:
I noticed that these functions don't have a case where they return a non-nil error. Do we have test coverage for what happens when the k8s calls are failing?

Reply:
🤔 I don't think we do very often, just because it ties into the workqueue, and the workerqueue retrying on error is based on the K8s libraries, which have their own tests and is pretty rock solid.

Any suggestion here though, I can see the point, just not quite sure how to tackle it effectively?

It's possible one could argue this would get picked up in e2e tests, which is a more dynamic environment, that does have errors that fire just because of the nature of k8s?

Reply:
That's what I was wondering - if this function returned an error once or twice and then the normal result would the test actually be any different? It sounds like no because the k8s libraries would just retry and cover up the errors with retries.